helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -463,6 +442,35 @@ public void onError(Exception e, ErrorCode code, ErrorType type) {
 
   }
 
+  // Verify the fromState and current state of the stateModel.
+  public Exception isMessageStaled(boolean inSchedulerCheck) {
+    String fromState = _message.getFromState();
+    String toState = _message.getToState();
+    String partitionName = _message.getPartitionName();
+
+    // state in _currentStateDelta uses current state from state model. It has the
+    // most up-to-date. current state. In case currentState in stateModel is null,
+    // partition is in initial state and we using it as current state.
+    // Defined in HelixStateMachineEngine.
+    String state = _currentStateDelta.getState(partitionName);
+
+    Exception err = null;
+    if (toState.equalsIgnoreCase(state)) {
+      // To state equals current state, we can just ignore the message
+      err = new HelixDuplicatedStateTransitionException(String
+          .format("Partition %s current state is same as toState (%s->%s) from message.",
+              partitionName, fromState, toState));
+    } else if (!inSchedulerCheck && fromState != null && !fromState.equals("*") && !fromState
```

Comment:
Besides the current_state equals to_state case, is this also a case that we can move to scheduling phase instead of execution phase? That could help reduce even more threads usage.

Reply:
TFTR. This case is a bit different than the above one. The handling for these two cases are different. 

For the first case when toState == currentState, we just treat the task as succeeded and finish task. There is no difference between doing the no-op when executing and bypass the task at scheduling phase.  

In the second case when currentState != fromState, executor will set the task as failed and report state transition error. Ir will also update ZK.

Reply:
I had an offline sync with Ali. He has submit a PR to fix this issue (#1390). I changed my PR accordingly. We now do the same validation for scheduling stage and execution stage.  
Currently my PR contains his patch. Will rebase after his PR merged.